File: src/tools/get-diff.ts
import { tool } from "ai";
import { runCommand } from "../run-command";
import { z } from "zod";
import { toAgentToolResult } from "../agent-tool-result";
import {
  executeToolWithState,
  type ExecutionTracker,
} from "../execution-state";

type DiffMode = "stat" | "name-only" | "full";

type ExecuteGit = (
  args: string[],
) => Promise<{
  stdout: string;
  stderr: string;
  exitCode: number;
}>;

const diffModeArgs: Record<DiffMode, string[]> = {
  stat: ["diff", "--stat"],
  "name-only": ["diff", "--name-only"],
  full: ["diff"],
};

export async function getDiff(
  input: { mode: DiffMode },
  executeGit: ExecuteGit = async (args) => {
    return await runCommand("git", args);
  },
) {
  const result = await executeGit(diffModeArgs[input.mode]);

  return {
    mode: input.mode,
    stdout: result.stdout,
    stderr: result.stderr,
    exitCode: result.exitCode,
  };
}

export function createGetDiffTool(options?: {
  executionTracker?: ExecutionTracker;
  executeGit?: ExecuteGit;
}) {
  return tool({
    description: "Show the current git diff in a read-only mode",

    inputSchema: z.object({
      mode: z.enum(["stat", "name-only", "full"]).default("stat"),
    }),

    execute: async ({ mode }) => {
      return await toAgentToolResult(async () =>
        await executeToolWithState({
          toolName: "getDiff",
          tracker: options?.executionTracker,
          run: async () => await getDiff({ mode }, options?.executeGit),
        })
      );
    },
  });
}

export const getDiffTool = createGetDiffTool();

File: internal/tools/getdiff.go
package tools

import (
	"context"
	"encoding/json"
	"fmt"
)

type getDiffTool struct{}

func (t *getDiffTool) Name() string        { return "getDiff" }
func (t *getDiffTool) Effect() Effect      { return EffectRead }
func (t *getDiffTool) Description() string { return "Show the current git diff" }
func (t *getDiffTool) Schema() map[string]any {
	return objectSchema(map[string]any{})
}

func (t *getDiffTool) Execute(ctx context.Context, input json.RawMessage) (any, error) {
	if err := decodeInput(input, &struct{}{}); err != nil {
		return nil, err
	}
	result, err := runCapturedCommand(ctx, "git", "diff")
	if err != nil {
		return nil, fmt.Errorf("git diff failed: %w\n%s", err, result.Stderr)
	}
	return result.Stdout, nil
}

func NewGetDiffTool() Tool { return &getDiffTool{} }
